float64 timing
float64 spacing